regression in 0.143.0: flag --allow-no-matching-release does not work as in version 0.142.0

example helmfile:
```
repositories:
  - name: mittwald
    url: https://helm.mittwald.de
releases:
  - name: secret-generator
    namespace: kube-system
    chart: mittwald/kubernetes-secret-generator
    installed: false
    labels:
      phase: delete
  - name: secret-generator
    namespace: kube-system
    chart: mittwald/kubernetes-secret-generator
    version: 3.3.4
    values:
      - resources:
          limits:
            cpu: 100m
            memory: 400Mi
          requests:
            cpu: 100m
            memory: 400Mi
    strategicMergePatches:
      - fix-locked-leader.yaml

```

used command:
```
helmfile --state-values-file=/path/config.yaml --file=/path/helmfile.yaml --allow-no-matching-release --selector phase=setup  sync --skip-crds
```

thrown exception:
```
panic: assertion error: release "kube-system/secret-generator" has never been selected. This shouldn't happen!

goroutine 1 [running]:
github.com/roboll/helmfile/pkg/app.(*App).getSelectedReleases(0xc0001990e0, 0xc00031e000, 0x37)
	/home/circleci/workspace/helmfile/pkg/app/app.go:1207 +0xebe
github.com/roboll/helmfile/pkg/app.(*App).sync(0xc0001990e0, 0xc00031e000, {0x1ae5e88, 0xc0002a9700})
	/home/circleci/workspace/helmfile/pkg/app/app.go:1645 +0x95
github.com/roboll/helmfile/pkg/app.(*App).Sync.func1.1()
	/home/circleci/workspace/helmfile/pkg/app/app.go:368 +0x3b
github.com/roboll/helmfile/pkg/app.(*Run).withPreparedCharts(0xc00031e000, {0x15c4d61, 0x4}, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000b54050, ...}, ...)
	/home/circleci/workspace/helmfile/pkg/app/run.go:91 +0x5ca
github.com/roboll/helmfile/pkg/app.(*App).Sync.func1(0xc00031e000)
	/home/circleci/workspace/helmfile/pkg/app/app.go:360 +0x234
github.com/roboll/helmfile/pkg/app.(*App).ForEachState.func1(0xc0005b4000)
	/home/circleci/workspace/helmfile/pkg/app/app.go:899 +0xef
github.com/roboll/helmfile/pkg/app.(*App).visitStates.func1({0x7ffc95adedb1, 0xd}, {0x7ffc95aded71, 0x3f})
	/home/circleci/workspace/helmfile/pkg/app/app.go:847 +0x722
github.com/roboll/helmfile/pkg/app.(*App).visitStateFiles.func1()
	/home/circleci/workspace/helmfile/pkg/app/app.go:679 +0x76
github.com/roboll/helmfile/pkg/app.(*App).within(0xc0001990e0, {0x7ffc95aded71, 0x3f}, 0xc00089d4a0)
	/home/circleci/workspace/helmfile/pkg/app/app.go:640 +0x267
github.com/roboll/helmfile/pkg/app.(*App).visitStateFiles(0xc0001990e0, {0x7ffc95aded71, 0xc00026fa38}, {{0xc0002a9290, 0x1, 0x1}, {{0xc0002a9720, 0x1, 0x1}}, 0x0, ...}, ...)
	/home/circleci/workspace/helmfile/pkg/app/app.go:673 +0x373
github.com/roboll/helmfile/pkg/app.(*App).visitStates(0xc0001990e0, {0x7ffc95aded71, 0x40cc94}, {{0xc0002a9290, 0x1, 0x1}, {{0xc0002a9720, 0x1, 0x1}}, 0x0, ...}, ...)
	/home/circleci/workspace/helmfile/pkg/app/app.go:759 +0xdd
github.com/roboll/helmfile/pkg/app.(*App).visitStatesWithSelectorsAndRemoteSupport(0xc0001990e0, {0x7ffc95aded71, 0x4d}, 0xc00089d760, 0x0, {0x0, 0x0, 0x15e1af9})
	/home/circleci/workspace/helmfile/pkg/app/app.go:1019 +0x4b2
github.com/roboll/helmfile/pkg/app.(*App).ForEachState(0xc0001990e0, 0xc00089d7c8, 0xe8, {0x0, 0x0, 0x0})
	/home/circleci/workspace/helmfile/pkg/app/app.go:895 +0xe5
github.com/roboll/helmfile/pkg/app.(*App).Sync(0xc0001990e0, {0x1ae5e88, 0xc0002a9700})
	/home/circleci/workspace/helmfile/pkg/app/app.go:357 +0x8a
main.main.func9(0x1ae5c98, {0xc00003cc60, 0x0})
	/home/circleci/workspace/helmfile/main.go:441 +0x4e
main.action.func1(0xc00003cc60)
	/home/circleci/workspace/helmfile/main.go:961 +0x86
github.com/urfave/cli.HandleAction({0x132e040, 0xc0002a8c60}, 0x4)
	/home/circleci/workspace/helmfile/vendor/github.com/urfave/cli/app.go:524 +0xa8
github.com/urfave/cli.Command.Run({{0x15c4d61, 0x4}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x16175c1, 0x43}, {0x0, ...}, ...}, ...)
	/home/circleci/workspace/helmfile/vendor/github.com/urfave/cli/command.go:173 +0x652
github.com/urfave/cli.(*App).Run(0xc0003836c0, {0xc00013a000, 0x8, 0x8})
	/home/circleci/workspace/helmfile/vendor/github.com/urfave/cli/app.go:277 +0x705
main.main()
	/home/circleci/workspace/helmfile/main.go:692 +0x5005
```

--- a/pkg/app/app.go
+++ b/pkg/app/app.go
@@ -1203,13 +1203,11 @@ func (a *App) getSelectedReleases(r *Run, includeTransitiveNeeds bool) ([]state.
 		// releases in the helmfile config.
 		// Otherwise we can't compute the DAG of releases correctly.
 		r, deduped := selectedIds[id]
-		if !deduped {
-			panic(fmt.Errorf("assertion error: release %q has never been selected. This shouldn't happen!", id))
+		if deduped {
+			deduplicated = append(deduplicated, r)
+
+			dedupedBefore[id] = struct{}{}
 		}
-
-		deduplicated = append(deduplicated, r)
-
-		dedupedBefore[id] = struct{}{}
 	}
 
 	if err := checkDuplicates(r.helm, r.state, deduplicated); err != nil {

--- a/pkg/app/app_apply_test.go
+++ b/pkg/app/app_apply_test.go
@@ -1220,4 +1220,65 @@ releases:
 			concurrency: 1,
 		})
 	})
+
+	t.Run("select non existant release with --allow-no-matching-release", func(t *testing.T) {
+		check(t, testcase{
+			files: map[string]string{
+				"/path/to/helmfile.yaml": `
+releases:
+- name: foo
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+
+- name: bar
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+`,
+			},
+			selectors: []string{"app=foo"},
+			upgraded:  []exectest.Release{},
+			error:     "err: no releases found that matches specified selector(app=foo) and environment(default), in any helmfile",
+			// as we check for log output, set concurrency to 1 to avoid non-deterministic test result
+			concurrency: 1,
+		})
+	})
+
+	t.Run("select single release from helmfile with two duplicates", func(t *testing.T) {
+		check(t, testcase{
+			files: map[string]string{
+				"/path/to/helmfile.yaml": `
+releases:
+- name: foo
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+
+- name: bar
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: build
+
+- name: bar
+  chart: incubator/raw
+  namespace: default
+  labels:
+    app: test
+`,
+			},
+			selectors: []string{"name=foo"},
+			upgraded:  []exectest.Release{},
+			diffs: map[exectest.DiffKey]error{
+				exectest.DiffKey{Name: "foo", Chart: "incubator/raw", Flags: "--kube-contextdefault--namespacedefault--detailed-exitcode"}: helmexec.ExitError{Code: 2},
+			},
+			error: "",
+			// as we check for log output, set concurrency to 1 to avoid non-deterministic test result
+			concurrency: 1,
+		})
+	})
 }

--- a/pkg/app/testdata/testapply_2/select_non_existant_release_with_--allow-no-matching-release/log
+++ b/pkg/app/testdata/testapply_2/select_non_existant_release_with_--allow-no-matching-release/log
@@ -1,0 +1,43 @@
+processing file "helmfile.yaml" in directory "."
+first-pass rendering starting for "helmfile.yaml.part.0": inherited=&{default map[] map[]}, overrode=<nil>
+first-pass uses: &{default map[] map[]}
+first-pass rendering output of "helmfile.yaml.part.0":
+ 0: 
+ 1: releases:
+ 2: - name: foo
+ 3:   chart: incubator/raw
+ 4:   namespace: default
+ 5:   labels:
+ 6:     app: test
+ 7: 
+ 8: - name: bar
+ 9:   chart: incubator/raw
+10:   namespace: default
+11:   labels:
+12:     app: test
+13: 
+
+first-pass produced: &{default map[] map[]}
+first-pass rendering result of "helmfile.yaml.part.0": {default map[] map[]}
+vals:
+map[]
+defaultVals:[]
+second-pass rendering result of "helmfile.yaml.part.0":
+ 0: 
+ 1: releases:
+ 2: - name: foo
+ 3:   chart: incubator/raw
+ 4:   namespace: default
+ 5:   labels:
+ 6:     app: test
+ 7: 
+ 8: - name: bar
+ 9:   chart: incubator/raw
+10:   namespace: default
+11:   labels:
+12:     app: test
+13: 
+
+merged environment: &{default map[] map[]}
+0 release(s) matching app=foo found in helmfile.yaml
+

--- a/pkg/app/testdata/testapply_2/select_single_release_from_helmfile_with_two_duplicates/log
+++ b/pkg/app/testdata/testapply_2/select_single_release_from_helmfile_with_two_duplicates/log
@@ -1,0 +1,69 @@
+processing file "helmfile.yaml" in directory "."
+first-pass rendering starting for "helmfile.yaml.part.0": inherited=&{default map[] map[]}, overrode=<nil>
+first-pass uses: &{default map[] map[]}
+first-pass rendering output of "helmfile.yaml.part.0":
+ 0: 
+ 1: releases:
+ 2: - name: foo
+ 3:   chart: incubator/raw
+ 4:   namespace: default
+ 5:   labels:
+ 6:     app: test
+ 7: 
+ 8: - name: bar
+ 9:   chart: incubator/raw
+10:   namespace: default
+11:   labels:
+12:     app: build
+13: 
+14: - name: bar
+15:   chart: incubator/raw
+16:   namespace: default
+17:   labels:
+18:     app: test
+19: 
+
+first-pass produced: &{default map[] map[]}
+first-pass rendering result of "helmfile.yaml.part.0": {default map[] map[]}
+vals:
+map[]
+defaultVals:[]
+second-pass rendering result of "helmfile.yaml.part.0":
+ 0: 
+ 1: releases:
+ 2: - name: foo
+ 3:   chart: incubator/raw
+ 4:   namespace: default
+ 5:   labels:
+ 6:     app: test
+ 7: 
+ 8: - name: bar
+ 9:   chart: incubator/raw
+10:   namespace: default
+11:   labels:
+12:     app: build
+13: 
+14: - name: bar
+15:   chart: incubator/raw
+16:   namespace: default
+17:   labels:
+18:     app: test
+19: 
+
+merged environment: &{default map[] map[]}
+1 release(s) matching name=foo found in helmfile.yaml
+
+Affected releases are:
+  foo (incubator/raw) UPDATED
+
+processing 1 groups of releases in this order:
+GROUP RELEASES
+1     default/default/foo
+
+processing releases in group 1/1: default/default/foo
+getting deployed release version failed:unexpected list key: {^foo$ --kube-contextdefault--deleting--deployed--failed--pending}
+
+UPDATED RELEASES:
+NAME   CHART           VERSION
+foo    incubator/raw          
+

